### PR TITLE
[WIP] Move Console usage of stat wrapper to new IO shim

### PR DESCRIPTION
**Pull request by nguerrera**
# WIP

Managed half of https://github.com/dotnet/coreclr/pull/1225 to get some feedback on the approach I'm leaning towards to address #2137.

See notes there. I've started to add some guidelines to the docs as well.

### CODE REVIEW

**Comment by nguerrera** [comment]
cc @stephentoub @ellismg @jkotas

**Comment by nguerrera** [comment]
Note that this can't pass CI without dotnet/coreclr#1225

File: Documentation/coding-guidelines/interop-guidelines.md [hunk 1]
```diff
@@ -244,3 +249,42 @@ rather than the underlying integral type.
     should be named with the most discoverable name possible; if that name
     is a concept (e.g. Errors), it can be named using managed naming
     guidelines.
+
+
+## UNIX shims
+
+Often, various UNIX flavors offer the same API from the point-of-view of compatibility
+with C/C++ source code, but they do not have the same ABI. e.g. Fields can be laid out
+differently, constants can have different numeric values, etc.
```
Comment (stephentoub) [inline Documentation/coding-guidelines/interop-guidelines.md]: Another big one is that exports can be named differently, making the DllImport mechanism itself quite challenging.

File: Documentation/coding-guidelines/interop-guidelines.md [hunk 2]
```diff
@@ -244,3 +249,42 @@ rather than the underlying integral type.
...
+
+This leaves us with a situation where we can't write portable P/Invoke declarations
+that will work on all flavors, and writing separate declarations per flavor won't scale.
```
Comment (stephentoub) [inline Documentation/coding-guidelines/interop-guidelines.md]: And is quite fragile.

File: Documentation/coding-guidelines/interop-guidelines.md [hunk 3]
```diff
@@ -244,3 +249,42 @@ rather than the underlying integral type.
...
+
+To address this, we're moving to a model where all UNIX interop from corefx starts with 
+a P/Invoke to a C++ lib written specifically for corefx. These libs -- libcorefx_*.so 
+(aka "shims") -- are  intended to be very thin layer over underlying platform libraries.
```
Comment (stephentoub) [inline Documentation/coding-guidelines/interop-guidelines.md]: Nit: "thin layer" => "thin layers"

File: src/Common/src/Interop/Unix/libcorefx_io/Interop.libcorefx_io.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libcorefx_io/Interop.libcorefx_io.cs]: Why do we have both of these Interop.libcorefx_io.cs and Interop.GetFileStat.cs files?

Comment (nguerrera) [inline src/Common/src/Interop/Unix/libcorefx_io/Interop.libcorefx_io.cs]: Oops.

Comment (stephentoub) [inline src/Common/src/Interop/Unix/libcorefx_io/Interop.libcorefx_io.cs]: :)

**Comment by nguerrera** [comment]
Replacing this with a new pull request.

**Pull request closed without merging**